### PR TITLE
[V1][Spec Decode] Change Spec Decode Rejection Sampling API

### PR DESCRIPTION
**Pull request by LiuXiaoxuanPKU**
This PR tries to change the API of rejection sampler of speculative decoding. This is the preliminary step to support random sampling.

Tasks outside the scope of this PR.
1. write `draft_token_ids` to persistent batch, which can avoid list operations to convert tensor.
2. implement random sampling [here](https://github.com/LiuXiaoxuanPKU/vllm/blob/18a18153e933e8f607a69a0f881d9998ff16389c/vllm/v1/sample/rejection_sampler.py#L146) and [here](https://github.com/LiuXiaoxuanPKU/vllm/blob/18a18153e933e8f607a69a0f881d9998ff16389c/vllm/v1/sample/rejection_sampler.py#L92). Rejection sampling guarantees that the output distribution with spec decode matches that without spec decode. The drafting method/distribution will not affect the correctness, but only affect the token acceptance rate.

### CODE REVIEW

File: vllm/v1/worker/gpu_model_runner.py [hunk 1]
```diff
@@ -953,12 +953,23 @@ def execute_model(
...
+        else:
+            target_probs = self.model.sampler.compute_probs(
+                logits, sampling_metadata)
+            scheduled_request_ids = scheduler_output.num_scheduled_tokens.keys(
+            )
+            draft_token_ids = [
+                scheduler_output.scheduled_spec_decode_tokens.get(req_id, [])
+                for req_id in scheduled_request_ids
+            ]
+            sampler_output = self.model.sampler.rejection_sampler(
```
Comment (CXIAAAAA) [inline vllm/v1/worker/gpu_model_runner.py:971]: since rejection_sampler is removed from sampler, we could not access it anymore?

Comment (LiuXiaoxuanPKU) [reply vllm/v1/worker/gpu_model_runner.py]: yes, fixed! Thanks for pointing it out

File: vllm/v1/sample/sampler.py [hunk 2]
```diff
@@ -127,6 +116,14 @@ def sample(
         )
         return sampled
 
+    def compute_probs(self, logits: torch.Tensor,
+                      sampling_metadata: SamplingMetadata) -> torch.Tensor:
+        if sampling_metadata.all_greedy:
+            return logits
+        # Apply temperature.
```
Comment (WoosukKwon) [inline vllm/v1/sample/sampler.py:123]: Can you please add a comment that this is an in-place op changing the logits?

**Review by WoosukKwon — APPROVED** [review]
LGTM!
